echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/artifacts/ArtifactMatcher.java

```diff
@@ -86,6 +93,55 @@ public static boolean isConstraintInPayload(final Map constraints, final Map pay
     return true;
   }
 
+  /**
+   * Check that there is a key in the payload for each constraint declared in a Trigger. Also check
+   * that if there is a value for a given key, that the value matches the value in the payload.
+   *
+   * <p>The constraint key may accept a JsonPath expression for deeper json search, the evaluation
+   * value should be a String or a List<String>.
+   *
+   * @param constraints A map of constraints configured in the Trigger (eg, created in Deck). A
+   *     constraint is a [key, java regex value] pair or a [JsonPathExp, java regex value].
+   * @param payload A map of the payload contents POST'd in the triggering event.
+   * @return Whether every key or expression (and value if applicable) in the constraints map is
+   *     represented in the payload.
+   */
+  public static boolean isJsonPathConstraintInPayload(final Map constraints, final Map payload) {
+    String json = gson.toJson(payload);
+    DocumentContext documentContext = JsonPath.using(conf).parse(json);
+
+    for (Object key : constraints.keySet()) {
+      if (!payload.containsKey(key) || payload.get(key) == null) {
+        String value;
+        try {
+          value = documentContext.read(key.toString(), String.class);
+        } catch (InvalidPathException e) {
+          return false;
+        }
+        if (value != null) {
+          return matches(constraints.get(key).toString(), value);
+        } else {
+          List<String> values;
+          try {
+            values = documentContext.read(key.toString());
+          } catch (ClassCastException e) {
+            return false;
+          }
+          if (values != null) {
+            return values.stream().anyMatch(v -> matches(constraints.get(key).toString(), v));
```

Comment:
This is a little unexpected to me but I can't exactly put my finger on why it's strange. I'm not sure that this is what I would expect if I wrote jsonpath that results in an array...

Reply:
I changed the implementation for this, but basically if any string in the list match the regex it is considered a valid match.